src/server/pfs/server/driver_new_storage.go

```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
 	return nil
 }
 
-func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes []string) (retErr error) {
+// compact is the entrypoint to compaction.
+func (d *driver) compact(ctx context.Context, master *work.Master, outputPath string, inputPrefixes []string) error {
+	// resolve prefixes into paths
+	inputPaths := []string{}
+	for _, inputPrefix := range inputPrefixes {
+		err := d.storage.WalkFileSets(ctx, inputPrefix, func(inputPath string) error {
+			inputPaths = append(inputPaths, inputPath)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	if d.env.StorageCompactionMaxFanIn < 2 {
+		panic("StorageCompactionMaxFanIn cannot be < 2")
+	}
+	return d.compactIter(ctx, compactParams{
+		master:     master,
+		inputPaths: inputPaths,
+		outputPath: outputPath,
+		maxFanIn:   d.env.StorageCompactionMaxFanIn,
+	})
+}
+
+type compactParams struct {
+	master     *work.Master
+	outputPath string
+	inputPaths []string
+	maxFanIn   int
+}
+
+// compactIter is one level of compaction.  It will only perform compaction
+// if len(inputPaths) < params.maxFanIn otherwise it will split inputPaths recursively.
+func (d *driver) compactIter(ctx context.Context, params compactParams) (retErr error) {
+	scratch := path.Join(tmpPrefix, uuid.NewWithoutDashes())
+	defer func() {
+		if err := d.storage.Delete(ctx, scratch); retErr == nil {
+			retErr = err
+		}
+	}()
+	// exceeds maxFanIn need to branch out
+	inputPaths := params.inputPaths
+	if len(inputPaths) > params.maxFanIn {
+		childOutputPaths := []string{}
+		childSize := len(params.inputPaths) / params.maxFanIn
+		if len(params.inputPaths)%params.maxFanIn != 0 {
+			childSize++
+		}
+		group, ctx := errgroup.WithContext(ctx)
+		for i := 0; i < params.maxFanIn; i++ {
+			start := i * childSize
+			end := (i + 1) * childSize
+			if end > len(inputPaths) {
+				end = len(inputPaths)
+			}
+			childOutputPath := path.Join(scratch, strconv.Itoa(i))
+			childOutputPaths = append(childOutputPaths, childOutputPath)
+			group.Go(func() error {
+				return d.compactIter(ctx, compactParams{
+					master:     params.master,
+					inputPaths: inputPaths[start:end],
+					outputPath: childOutputPath,
+					maxFanIn:   params.maxFanIn,
+				})
+			})
+		}
+		if err := group.Wait(); err != nil {
+			return err
+		}
+		inputPaths = childOutputPaths
+	}
+	// within maxFanIn, generate compaction tasks for workers
+	return d.shardedCompact(ctx, params.master, params.outputPath, inputPaths)
+}
+
+// shardedCompact generates shards for the fileset(s) in inputPaths,
+// gives those shards to workers, and waits for them to complete
+// fan-in: spatially, and temporally to each shard will be len(inputPaths)
```

Comment:
Not sure I understand what these two lines are saying? Maybe reword? Fan in is bound by len(inputPaths), concatenating shards have fan in of one because they are concatenated sequentially.